### ISSUE TITLE
Automatic snapshot on alarm

### ISSUE THREAD

**Issue by ririsoft** [netdata/netdata#3973]
Hello,

Since this is my first issue here, I would like to congratulate all the people working on Netdata. It is a brilliant project with a clear vision. We see behind such tool the solid DevOps experience from the developers. **Thank you very much for your work !**

I would like to discuss with you the possibility to have **Automatic Snapshot on Alarm** as an alternative to a longer history storage as discussed in #3333 for instance. The user stories are the following :

1. DevOp is in a meeting with customers. He receives a warning alarm which quickly recovers. However DevOp would like to investigate the issue later on after his meeting. He his about to present the brand new product's architecture to the customer for 30 minutes, he does not have the time to go the Netadata site and take a snapshot.

2. Netdata has been configured to not send mail/SMS to DevOp for warnings which quickly recover. When DevOp arrives at work in the morning he sees the alarm history and would like to investigate what happened during the night. 

In such situations (and most alarm situations ?) **snapshots are perfectly fine**. When an issue comes in i only want to know what happened a few minutes before and how the system reacts a few minutes after. I do not need to have a full history for days or months. For big data / analytics I would rather use Prometheus, Elastic Search, Hadoop or other. But for real time monitoring Netdata snapshots are perfect. Also having a lower sampling rate for longer history does not help here. I really want the true snapshot at 1 second resolution like if I were available at the time the alarm raises.

**However when an issue arrives i might not be able to secure the snapshot** for the past hour and next 30 minutes (in a meeting, at night, ...). It would be great if **Netdata takes the snapshot for me automatically**.

I believe I can do it already with a custom alarm-notify script and the Netdata API (i did not investigated this further), but having that in Netdata natively would be awesome.

Thinking at how it could be implemented I have been through 3 ideas:

1. Take the snapshot automatically and store it on disk. Such storage has to be limited in size and time. Such snapshots could be accessible through the web UI.

2. Attach the snapshot to the mail (slack, hipchat, ...) notification.

3. Send the snapshot to another server (which protocol ?).

What do you think ? Couldn't it be an alternative to a longer history ?
Cheers,
Riri.

**Comment by Ferroin** [comment]
So, first off, this is an awesome idea, and I think it's a great alternative option to having a significantly extended history.  I think you are correct that it could be done with a custom alarm-notify script, but it would indeed be better for usability to just have it natively integrated.

Personally, I'd vote for the first idea you have for implementation.  If that approach is taken, the links sent in notifications could optionally be changed to point to the snapshot instead of the live chart, which would allow for a simplified version of the second idea.

**Comment by ririsoft** [comment]
> Personally, I'd vote for the first idea you have for implementation. If that approach is taken, the links sent in notifications could optionally be changed to point to the snapshot instead of the live chart, which would allow for a simplified version of the second idea.

I fully agree, this would be the best user experience. Attaching to the mail would have the advantage to secure the snapshot in case of disk crash or system unavailability, but in that case we all agree that it is not what mails are for and that the solution 3 is the way to go.

**Comment by Ferroin** [comment]
Solution 3 could be as simple as just letting a Netdata master node handle it.

**Comment by cakrit** [comment]
@gmosx I believe that if we can combine this with #3643, it will be a killer for the hub. Where the snapshot will be stored and limiting the amount of storage required so we don't abuse the disk in an alarm storm is something that will need some design of course.

**Comment by cakrit** [comment]
My suggestion is the following:

1. Extend health/health.c to:
 - Export up to SNAPSHOTS_NFILES snapshots to disk (default 0)
 - Using compression SNAPSHOTS_COMPRESSION (possible values pako.deflate, pako.deflate.base64, lzstring.uri, lzstring.utf16, lzstring.base64) (default pako.deflate.base64). 
 - With SNAPSHOTS_SEC_PER_POINT seconds per point (default 1)
 - Keeping data for the last SNAPSHOTS_MINUTES_PER_SNAPSHOT minutes
 - Perhaps naming the file with the raised `alarm identifier`, to allow for predictable paths. 

The params will be configured in health_alarm_notify.conf
Files will be deleted FIFO, based on either timestamp, or looking up the alarm entries in the alarm log.

2. Modify the web server to:
 2.1 Accept URLs that specify that a snapshot for alarm `alarm identifier` should be automatically imported and displayed.
 2.2 Perhaps accept URLs that specify that a snapshot for alarm `alarm identifier` should be returned for download (not sure if it's required).

3. On the UI:
 3.1 Add a button to "Active Alarms" to go to the snapshot of that alarm (URL 2.1)
 3.2 Add buttons to "Alarm Log Entries" that do have snapshots, to go the snapshot of each alarm (URL 2.1). 
 Even if 2.2 is required, I don't think we want to put buttons for them on the UI.

4. Not sure if we want to send URL 2.1 and/or 2.2 to alarm-notify.sh. Open to suggestions.

Finally, the hub will probably do something different:
- Add the Authentication/Authorization layer
- Perhaps enable viewing the snapshots without the user having access to the netdata server URL?

**Comment by cakrit** [comment]
We have discussed with @ktsaou that the proposed solution would require great care in deciding when to take the snapshot, to guarantee that we don't stress the server when it has a problem. Several alternatives were discussed and we'll think about it a bit more before starting the work. Lowering priority to medium for now.

**Comment by ririsoft** [comment]
Since the introduction of the dbengine the need for snapshot has evolved: I now have enough retention and do not need snapshot storage anymore.

However I need to browse my alarm log history and be able to jump to the related time and chart even if the alarm is not active anymore. I do not know if I can do that we the current UI or if a change request is needed for that. It seems to be available only for active alarms.

Would you prefer that I open a separate issue for that ?

**Comment by cakrit** [comment]
You're correct @ririsoft , the PR for #5810 is #6391 and I'm sure it will make it into v1.17.
You're making a very interesting point about the usefulness of snapshots, given a longer-term retention period. It's something we'll definitely need to reconsider, if we don't find other use cases for this feature..

**Comment by vlvkobal** [comment]
Taking into account that
> I now have enough retention and do not need snapshot storage anymore.

and #6391 was merged I'm closing the issue.
